src/datatypes/mod.rs

```diff
@@ -160,6 +160,126 @@ pub enum DataType {
     Extension(String, Box<DataType>, Option<String>),
 }
 
+#[cfg(feature = "arrow")]
+impl From<DataType> for arrow_schema::DataType {
+    fn from(value: DataType) -> Self {
+        match value {
+            DataType::Null => Self::Null,
+            DataType::Boolean => Self::Boolean,
+            DataType::Int8 => Self::Int8,
+            DataType::Int16 => Self::Int16,
+            DataType::Int32 => Self::Int32,
+            DataType::Int64 => Self::Int64,
+            DataType::UInt8 => Self::UInt8,
+            DataType::UInt16 => Self::UInt16,
+            DataType::UInt32 => Self::UInt32,
+            DataType::UInt64 => Self::UInt64,
+            DataType::Float16 => Self::Float16,
+            DataType::Float32 => Self::Float32,
+            DataType::Float64 => Self::Float64,
+            DataType::Timestamp(unit, tz) => Self::Timestamp(unit.into(), tz),
+            DataType::Date32 => Self::Date32,
+            DataType::Date64 => Self::Date64,
+            DataType::Time32(unit) => Self::Time32(unit.into()),
+            DataType::Time64(unit) => Self::Time64(unit.into()),
+            DataType::Duration(unit) => Self::Duration(unit.into()),
+            DataType::Interval(unit) => Self::Interval(unit.into()),
+            DataType::Binary => Self::Binary,
+            DataType::FixedSizeBinary(size) => Self::FixedSizeBinary(size as _),
+            DataType::LargeBinary => Self::LargeBinary,
+            DataType::Utf8 => Self::Utf8,
+            DataType::LargeUtf8 => Self::LargeUtf8,
+            DataType::List(f) => Self::List(Box::new((*f).into())),
+            DataType::FixedSizeList(f, size) => {
+                Self::FixedSizeList(Box::new((*f).into()), size as _)
+            }
+            DataType::LargeList(f) => Self::LargeList(Box::new((*f).into())),
+            DataType::Struct(f) => Self::Struct(f.into_iter().map(Into::into).collect()),
+            DataType::Union(fields, Some(ids), mode) => {
+                let ids = ids.into_iter().map(|x| x as _).collect();
+                let fields = fields.into_iter().map(Into::into).collect();
+                Self::Union(fields, ids, mode.into())
+            }
+            DataType::Union(fields, None, mode) => {
+                let ids = (0..fields.len() as i8).collect();
+                let fields = fields.into_iter().map(Into::into).collect();
+                Self::Union(fields, ids, mode.into())
+            }
+            DataType::Map(f, ordered) => Self::Map(Box::new((*f).into()), ordered),
+            DataType::Dictionary(key, value, _) => Self::Dictionary(
+                Box::new(DataType::from(key).into()),
+                Box::new((*value).into()),
+            ),
+            DataType::Decimal(precision, scale) => Self::Decimal128(precision as _, scale as _),
+            DataType::Decimal256(precision, scale) => Self::Decimal256(precision as _, scale as _),
+            DataType::Extension(_, d, _) => (*d).into(),
```

Comment:
I'm curious why arrow2 chose to represent extension types as an explicit data type, as opposed to just field metadata?

Reply:
It is so that the `Array` has the logical type on it. It allows the user of a `Box<dyn Array>` to use `.data_type()` to perform the downcast and have the necessary information to build the extension.

For example, polars uses it to store arbitrary Python objects on the type. In theory this could be kept in the Field's metadata.

Pyarrow does the same: https://arrow.apache.org/docs/python/generated/pyarrow.ExtensionType.html